### PR TITLE
Use python and python libraries from CVMFS COMP under slc7_amd64_gcc630

### LINKED ISSUE #9817: WMAgent rc 1.3.6.patch1 fails during runtime when importing future
**Impact of the bug**
WMAgent

**Describe the bug**
This patch version is the first release candidate for WMAgent stable release 1.3.6
It contains a few of the python3 compatibility changes, and I noticed that those py3 changes cause 100% job failure rate.

The error happens in the worker nodes during runtime, when importing `future`, such as:
```
    from future.utils import viewitems
ImportError: No module named future.utils
```

**How to reproduce it**
Not sure, if I source the agent environment, I can successfuly import from future:
```
cmst1@vocms0192:/data/srv/wmagent/current $ source apps/wmagent/etc/profile.d/init.sh 
cmst1@vocms0192:/data/srv/wmagent/current $ python
Python 2.7.13+ (default, Dec 18 2018, 10:09:09) 
[GCC 6.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from future.utils import viewitems
>>
```

**Expected behavior**
Future should be available with the COMP python environment, which is meant to be load/used during runtime.

**Additional context and error message**
From the testbed campaign `Agent136_Val`:
```
NoJobReport (Exit Code: 99303)

    Could not find jobReport

    ========== condor.14266.32.err ==========
    Adding end of condor.14266.32.err to error message:

    WARNING: container does not have /.singularity.d/actions/exec, calling /srv/.gwms-user-job-wrapper.sh directly
    WARNING  glidein_config not defined () in singularity_lib.sh. Some functions like advertise and error_gen will be limited.
    Traceback (most recent call last):
      File "Startup.py", line 17, in <module>
        import WMCore.WMRuntime.Bootstrap as Bootstrap
      File "/srv/job/WMCore.zip/WMCore/WMRuntime/Bootstrap.py", line 23, in <module>
      File "/srv/job/WMCore.zip/WMCore/WMRuntime/TaskSpace.py", line 16, in <module>
      File "/srv/job/WMCore.zip/WMCore/WMSpec/WMWorkload.py", line 10, in <module>
      File "/srv/job/Utils/Utilities.py", line 4, in <module>
        from future.utils import viewitems
    ImportError: No module named future.utils
    + cp 'WMTaskSpace/Report*.pkl' ../
    cp: cannot stat `WMTaskSpace/Report*.pkl': No such file or directory
    + ls -l WMTaskSpace
    ls: cannot access WMTaskSpace: No such file or directory
    + ls -l 'WMTaskSpace/*'
    ls: cannot access WMTaskSpace/*: No such file or directory
    + set +x
```

From the _condor.out (which makes me wonder, why are we still using the old scram arch...)
```
...
======== WMAgent Python boostrap starting at Thu Jul  9 10:06:34 GMT 2020 ========
WMAgent bootstrap: latest python release is: 2.7.13
WMAgent bootstrap: found python2 at..
/cvmfs/cms.cern.ch/COMP/slc6_amd64_gcc493/external/python/2.7.13/bin/python2
======== WMAgent Python bootstrap finished at Thu Jul  9 10:06:34 GMT 2020 ========
...
  PYTHONPATH=/cvmfs/oasis.opensciencegrid.org/mis/osg-wn-client/3.4/3.4.53/el6-x86_64/usr/lib/python2.6/site-packages:/cvmfs/oasis.opensciencegrid.org/mis/osg-wn-client/3.4/3.4.53/el6-x86_64/usr/lib64/python2.6/site-packages:/srv/job/WMCore.zip:/srv/job 
```

### LINKED ISSUE #9876: PY3: Jobs cannot import from future with WMAgent 1.3.6.patch4 and WMCore 1.4.0.pre5
**Impact of the bug**
WMCore in general

**Describe the bug**
WMCore central services have just been deployed in testbed, with 1.4.0.pre5. Services seem to be running fine so far, but workflows are hitting 100% failure rate. They fail on the worker node when the job wrapper is loading the workload spec file, traceback at the bottom.

**How to reproduce it**
Try to unpickle that object with the agent environment, but without py2-future in the PYTHONPATH.

This works fine within the agent - as expected, because it has the future package in the path.
```
cmst1@vocms0193:/data/srv/wmagent/current $ source apps/wmagent/etc/profile.d/init.sh 
cmst1@vocms0193:/data/srv/wmagent/current $ python
Python 2.7.13+ (default, Dec 18 2018, 10:09:09) 
[GCC 6.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import pickle
>>> workloadPcl = "/data/srv/wmagent/v1.3.6.patch4/install/wmagent/WorkQueueManager/cache/amaltaro_TaskChain_NanoAOD_HG2009_Val_200820_105910_758/WMSandbox/WMWorkload.pkl"
>>> with open(workloadPcl, 'r') as handle:
...   wmWorkload = pickle.load(handle)
... 
>>> wmWorkload
<WMCore.WMSpec.WMWorkload.WMWorkload object at 0x7fd0adc18f90>
```

**Expected behavior**
WMAgent 1.3.6.patch4 should be able to run these jobs without any issues, given that that version does not have the futurize stage2 fixers. 
We need to check how exactly the workload spec file changed, and see what are the changes needed on the agents to make it working.

Another very good alternative would be to actually sort out the py2-future dependency on the worker nodes (cvmfs), reported here:
https://github.com/dmwm/WMCore/issues/9817
Once that gets fixed, then all these futurize issues will be solved.

**Additional context and error message**
traceback from wmstats:
```
Job Name: 1c7fdcdf-88ef-4b01-bd27-b5ac6de91c37-0
WMBS job id: 3558
Workflow: amaltaro_TaskChain_NanoAOD_HG2009_Val_200820_105910_758
Task: /amaltaro_TaskChain_NanoAOD_HG2009_Val_200820_105910_758/RunIIFall17NanoAOD
Status: jobfailed
Input dataset: /WJetsToLNu_HT-200To400_TuneCUETP8M1_13TeV-madgraphMLM-pythia8/RunIISummer16MiniAODv2-PUMoriond17_80X_mcRun2_asymptotic_2016_TrancheIV_v6_ext1-v1/MINIAODSIM
Site: T2_CH_CERN
Agent: vocms0193.cern.ch
ACDC URL: https://cmsweb-testbed.cern.ch/couchdb/acdcserver
State Transition
Exit code: 99303
Retry count: 0
Errors:

    NoJobReport
        NoJobReport (Exit Code: 99303)

            Could not find jobReport

            ========== condor.14002.41.err ==========
            Adding end of condor.14002.41.err to error message:

            WARNING  glidein_config not defined () in singularity_lib.sh. Some functions like advertise and error_gen will be limited.
            INFO:root:Process id: 132	Current working directory: /srv/job
            INFO:root:Loading job definition
...
INFO:root:Loading task
Traceback (most recent call last):
  File "Startup.py", line 29, in <module>
    task = Bootstrap.loadTask(job)
  File "/srv/job/WMCore.zip/WMCore/WMRuntime/Bootstrap.py", line 185, in loadTask
  File "/srv/job/WMCore.zip/WMCore/WMRuntime/Bootstrap.py", line 172, in loadWorkload
  File "/cvmfs/cms.cern.ch/COMP/slc6_amd64_gcc493/external/python/2.7.13/lib/python2.7/pickle.py", line 1384, in load
    return Unpickler(file).load()
  File "/cvmfs/cms.cern.ch/COMP/slc6_amd64_gcc493/external/python/2.7.13/lib/python2.7/pickle.py", line 864, in load
    dispatch[key](self)
  File "/cvmfs/cms.cern.ch/COMP/slc6_amd64_gcc493/external/python/2.7.13/lib/python2.7/pickle.py", line 1096, in load_global
    klass = self.find_class(module, name)
  File "/cvmfs/cms.cern.ch/COMP/slc6_amd64_gcc493/external/python/2.7.13/lib/python2.7/pickle.py", line 1130, in find_class
    __import__(module)
ImportError: No module named future.types.newstr
+ cp 'WMTaskSpace/Report*.pkl' ../
cp: cannot stat `WMTaskSpace/Report*.pkl': No such file or directory
+ ls -l WMTaskSpace
ls: cannot access WMTaskSpace: No such file or directory
+ ls -l 'WMTaskSpace/*'
ls: cannot access WMTaskSpace/*: No such file or directory
+ set +x
```

### PR DESCRIPTION
**Pull request by amaltaro**
Fixes #9817 
Fixes #9876 

#### Status
ready

#### Description
Packages (python and py2-future) have been deployed into CVMFS (thanks to Shahzad!). With that available, we can update the job wrapper to source the correct ScramArch python build according to the job's REQUIRED_OS, such as:
* if requesting `rhel7`, then use `slc7_amd64_gcc630`
* else, keep using the old and known to work `slc6_amd64_gcc493` packages.

#### Is it backward compatible (if not, which system it affects?)
yes

#### Related PRs
none

#### External dependencies / deployment changes
cmsdist PR: https://github.com/cms-sw/cmsdist/pull/6056

### CODE REVIEW

**Comment by amaltaro** [comment]
With the current patch, plus the python + py2-future deployment in CVMFS for slc7_amd64_gcc630 (thanks to Shahzad!), this workflow still fails during the job bootstrap process:
Workflow: amaltaro_SC_ReDigi_Harvest_Agent136_Val_200711_082023_6067

with the following error:
```
2020-07-11 08:38:46,988:INFO:SetupCMSSWPset:Executing SetupCMSSWPSet...
2020-07-11 08:38:51,862:INFO:SetupCMSSWPset:Tag chirp updates from CMSSW with step cmsRun1
2020-07-11 08:38:51,863:INFO:WMTweak:Not resizing job
2020-07-11 08:38:51,864:INFO:WMTweak:Setting 'firstLuminosityBlock' attr to: 1
2020-07-11 08:38:51,864:INFO:WMTweak:Adding 0 files to 'fileNames' attr
2020-07-11 08:38:51,864:INFO:WMTweak:Adding 0 files to 'secondaryFileNames' attr
2020-07-11 08:38:51,864:INFO:WMTweak:Setting 'firstEvent' attr to: 1
2020-07-11 08:38:51,864:INFO:WMTweak:Setting 'maxEvents.input' attr to: 100
Traceback (most recent call last):
  File "/cvmfs/cms.cern.ch/slc7_amd64_gcc700/external/python/2.7.14-pafccj/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/cvmfs/cms.cern.ch/slc7_amd64_gcc700/external/python/2.7.14-pafccj/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/srv/job/WMCore.zip/WMCore/WMRuntime/ScriptInvoke.py", line 117, in <module>
RuntimeError: Error invoking script for step WMTaskSpace.cmsRun1
withe Script module: SetupCMSSWPset
FEVTDEBUGoutput.root is not a valid <class 'FWCore.ParameterSet.Types.string'>Details:
  File "/srv/job/WMCore.zip/WMCore/WMRuntime/ScriptInvoke.py", line 109, in <module>
  File "/srv/job/WMCore.zip/WMCore/WMRuntime/ScriptInvoke.py", line 80, in invoke
  File "/build/dmwmbld/srv/state/dmwmbld/builds/comp_gcc630/w/tmp/BUILDROOT/5d7f9928091f1a8936f62f30715c0a64/build/dmwmbld/srv/state/dmwmbld/builds/comp_gcc630/w/slc7_amd64_gcc630/cms/wmagent/1.3.6.patch1/lib/python2.7/site-packages/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py", line 798, in __call__
    applyTweak(self.process, outTweak, self.fixupDict)
  File "/srv/job/PSetTweaks/WMTweak.py", line 319, in applyTweak
    setParameter(process, param, value)
  File "/srv/job/PSetTweaks/WMTweak.py", line 199, in setParameter
    lastPSet.setValue(value)
  File "/cvmfs/cms.cern.ch/slc7_amd64_gcc700/cms/cmssw-patch/CMSSW_10_6_1_patch1/python/FWCore/ParameterSet/Mixins.py", line 97, in setValue
    raise ValueError(str(value)+" is not a valid "+str(type(self)))
```

**Comment by amaltaro** [comment]
@mapellidario Dario, as we discussed on Monday, this is my initial attempt to migrate the runtime code to use the python binary (and python-future) from
```
/cvmfs/cms.cern.ch/slc7_amd64_gcc700/external/python
```
instead of the current
```
/cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/python
```

**Comment by amaltaro** [comment]
After having all the necessary packages installed in CVMFS (python and py2-future) for both slc6 and slc7 architectures:
```
amaltaro@lxplus738:~ $ ls /cvmfs/cms.cern.ch/COMP/slc6_amd64_gcc493/external/
apt  bootstrap-bundle  bz2lib  db4  expat  gcc  gdbm  libevent  ncurses  openssl  py2-argparse  py2-future  py2-setuptools  python  readline  rpm  sqlite  xrootd  zlib

amaltaro@lxplus738:~ $ ls /cvmfs/cms.cern.ch/COMP/slc7_amd64_gcc630/external/
bootstrap-bundle  bz2lib  db4  expat  gcc  gdbm  ncurses  openssl  py2-future  py2-setuptools  python  readline  rpm  sqlite  zlib
```

and some updates to this patch, I can finally run the whole set of DMWM requests without any issues. Jobs were executed only at FNAL and CERN, and I believe we should test it in a few other sites before we roll it to production targetting the CMSWEB September upgrade.

**Review by khurtado — COMMENTED** [review]
Hi @amaltaro. Changes look good. I made a small suggestion and put a question at the end, but in general I think this is solid and safe to merge.

File: etc/submit.sh [hunk 1]
```diff
@@ -157,7 +168,7 @@ $pythonCommand Unpacker.py --sandbox=$SANDBOX --package=JobPackage.pkl --index=$
 
 cd job
 export WMAGENTJOBDIR=$PWD
-export PYTHONPATH=$PYTHONPATH:$PWD/WMCore.zip:$PWD
+export PYTHONPATH=$PYTHONPATH:$WMAGENTJOBDIR/WMCore.zip:$WMAGENTJOBDIR:$compPythonPath
```
[the remark this answers is not in the document]
Comment (amaltaro) [reply etc/submit.sh:171]: I thought we need the upper directory in the path, but you're probably right. Updating the code

File: etc/submit.sh [hunk 2]
```diff
@@ -131,11 +139,14 @@ else
...
 fi
 
+compPythonPath=`echo $prefix | sed 's|/python||'`
+echo "WMAgent bootstrap: COMP Python path is: $compPythonPath"
 latestPythonVersion=`ls -t "$prefix"/*/"$suffix" | head -n1 | sed 's|.*/external/python/||' | cut -d '/' -f1`
 pythonMajorVersion=`echo $latestPythonVersion | cut -d '.' -f1`
 pythonCommand="python"${pythonMajorVersion}
 echo "WMAgent bootstrap: latest python release is: $latestPythonVersion"
 source "$prefix"/"$latestPythonVersion"/"$suffix"
+source "$compPythonPath"/py2-future/0.18.2/"$suffix"
```
[the remark this answers is not in the document]
Comment (amaltaro) [reply etc/submit.sh:149]: Good idea! Update coming in a minute.

**Comment by amaltaro** [comment]
@khurtado Kenyi, the changes you suggested are provided in the second commit. If you're happy with those, I'll squash those commits.
Still need to test this new version though :-D

**Review by khurtado — APPROVED** [review]
@amaltaro : Everything looks good to me. Feel free to squash the commits

**Comment by amaltaro** [comment]
I ran all the Integration set of tests overnight - with the latest changes to this PR - and things are working nicely.

**Pull request merged**